samples/advanced/financial.html

```diff
@@ -0,0 +1,251 @@
+				return Math.random() * (max - min) + min;
+			}
+
+			function randomBar(date, lastClose) {
+				const open = randomNumber(lastClose * 0.95, lastClose * 1.05).toFixed(2);
+				const close = randomNumber(open * 0.95, open * 1.05).toFixed(2);
+				return {
+					t: date.valueOf(),
+					y: close
+				};
+			}
+
+			let date = moment('Jan 01 1990', 'MMM DD YYYY');
+			const now = moment();
+			const data = [];
+			const lessThanDay = unitLessThanDay();
+			for (; data.length < 600 && date.isBefore(now); date = date.clone().add(1, unit).startOf(unit)) {
+				if (outsideMarketHours(date)) {
+					if (!lessThanDay || !beforeNineThirty(date)) {
+						date = date.clone().add(date.isoWeekday() >= 5 ? 8 - date.isoWeekday() : 1, 'day');
+					}
+					if (lessThanDay) {
+						date = date.hour(9).minute(30).second(0);
+					}
+				}
+				data.push(randomBar(date, data.length > 0 ? data[data.length - 1].y : 30));
+			}
+
+			return data;
+		}
+
+		const ctx = document.getElementById('chart1').getContext('2d');
+		ctx.canvas.width = 1000;
+		ctx.canvas.height = 300;
+
+		const color = Chart.helpers.color;
+		const cfg = {
+			data: {
+				datasets: [{
+					label: 'CHRT - Chart.js Corporation',
+					backgroundColor: color(window.chartColors.red).alpha(0.5).rgbString(),
+					borderColor: window.chartColors.red,
+					data: generateData(),
+					type: 'line',
+					pointRadius: 0,
+					fill: false,
+					lineTension: 0,
+					borderWidth: 2
+				}]
+			},
+			options: {
+				animation: {
+					duration: 0
+				},
+				scales: {
+					x: {
+						type: 'time',
+						distribution: 'series',
+						offset: true,
+						ticks: {
+							major: {
+								enabled: true,
+							},
+							fontStyle: function(context) {
+								return context.tick.major ? 'bold' : undefined;
+							},
+							source: 'labels', // We provided no labels. Generate no ticks. We'll make our own
+							autoSkip: true,
+							autoSkipPadding: 75,
+							maxRotation: 0,
+							sampleSize: 100
+						},
+						// Custom logic that chooses ticks from dataset timestamp by choosing first timestamp in time period
+						afterBuildTicks: function(scale) {
+							// Determine units according to our own logic
+							// Make sure there's at least 10 ticks generated. autoSkip will remove any extras
+							const units = ['second', 'minute', 'hour', 'day', 'month', 'year'];
+							const adapter = scale._adapter;
+							const unit = document.getElementById('unit').value;
+							let minorUnit = unit;
+							for (let i = units.indexOf(minorUnit); i < units.length; i++) {
+								const periods = adapter.diff(scale.max, scale.min, units[i]);
+								if (periods < 10) {
+									break;
+								}
+								minorUnit = units[i];
+							}
+							let majorUnit;
+							if (units.indexOf(minorUnit) !== units.length - 1) {
+								majorUnit = units[units.indexOf(minorUnit) + 1];
+							}
+
+							// Generate ticks according to our own logic
+							const timestamps = scale._cache.all;
```

Comment:
Another private one. Could use `data` instead. (could use loopup on that data too, if lookup wasn't private)

Reply:
done